internal/servicelog/ringbuffer.go

```diff
@@ -0,0 +1,369 @@
+func (rb *RingBuffer) Buffered() int {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.buffered()
+}
+
+func (rb *RingBuffer) buffered() int {
+	return int(rb.writeIndex - rb.readIndex)
+}
+
+// Size returns the size in bytes of the internal buffer.
+func (rb *RingBuffer) Size() int {
+	return len(rb.data)
+}
+
+// Positions returns the start and end positions of readable data in the RingBuffer.
+func (rb *RingBuffer) Positions() (start RingPos, end RingPos) {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.readIndex, rb.writeIndex
+}
+
+// Copy copies bytes from the range into the supplied dest buffer. If dest is not large enough
+// to fill the bytes from start to end, then start to start+len(dest) is copied and
+// the error io.ErrShortBuffer is returned.
+func (rb *RingBuffer) Copy(dest []byte, start RingPos) (int, error) {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	if start < rb.readIndex || start > rb.writeIndex {
+		return 0, ErrRange
+	}
+	if rb.writeClosed && start == rb.writeIndex {
+		return 0, io.EOF
+	}
+	copyLength := int(rb.writeIndex - start)
+	if copyLength > len(dest) {
+		copyLength = len(dest)
+	}
+	if copyLength == 0 {
+		return 0, nil
+	}
+	end := start + RingPos(copyLength)
+	buffers := rb.buffers(start, end)
+	written := 0
+	for _, buffer := range buffers {
+		if len(buffer) == 0 {
+			continue
+		}
+		n := copy(dest, buffer)
+		dest = dest[n:]
+		written += n
+	}
+	if rb.writeClosed && start+RingPos(written) == rb.writeIndex {
+		return written, io.EOF
+	}
+	return written, nil
+}
+
+// WriteTo writes the selected range to a io.Writer.
+func (rb *RingBuffer) WriteTo(writer io.Writer, start RingPos) (int64, error) {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	if start < rb.readIndex || start > rb.writeIndex {
+		return 0, ErrRange
+	}
+	if rb.writeClosed && start == rb.writeIndex {
+		return 0, io.EOF
+	}
+	copyLength := rb.writeIndex - start
+	if copyLength == 0 {
+		return 0, nil
+	}
+	end := rb.writeIndex
+	buffers := rb.buffers(start, end)
+	written := int64(0)
+	for _, buffer := range buffers {
+		if len(buffer) == 0 {
+			continue
+		}
+		n, err := writer.Write(buffer)
+		written += int64(n)
+		if err != nil {
+			return written, err
+		}
+	}
+	if rb.writeClosed && start+RingPos(written) == rb.writeIndex {
+		return written, io.EOF
+	}
+	return written, nil
+}
+
+// TailIterator returns an iterator from the tail of the buffer.
+func (rb *RingBuffer) TailIterator() Iterator {
+	rb.iteratorMutex.Lock()
+	defer rb.iteratorMutex.Unlock()
+	start, _ := rb.Positions()
+	iter := &iterator{
+		rb:         rb,
+		index:      start,
+		notifyChan: make(chan struct{}),
+		closeChan:  make(chan struct{}),
+	}
+	if rb.Closed() {
+		close(iter.closeChan)
+	}
+	rb.iteratorList = append(rb.iteratorList, iter)
+	return iter
+}
+
+// HeadIterator returns an iterator from the head of the buffer.
+func (rb *RingBuffer) HeadIterator(lines int) Iterator {
```

Comment:
```
// If lines is greater than zero, the iterator will start that many lines
// backwards from the head.
```
... I suppose.

Reply:
That is a much better comment, thank-you.